crates/file_store/src/store.rs

```diff
@@ -135,23 +156,32 @@ where
     /// **WARNING**: This method changes the write position of the underlying file. The next
     /// changeset will be written over the erroring entry (or the end of the file if none existed).
     pub fn aggregate_changesets(&mut self) -> Result<Option<C>, AggregateChangesetsError<C>> {
-        let mut changeset = Option::<C>::None;
-        for next_changeset in self.iter_changesets() {
-            let next_changeset = match next_changeset {
-                Ok(next_changeset) => next_changeset,
-                Err(iter_error) => {
-                    return Err(AggregateChangesetsError {
-                        changeset,
-                        iter_error,
-                    })
+        let changeset_aggregator =
+            |mut aggregated_changeset: Option<C>, next_changeset| match next_changeset {
+                Ok(next_changeset) => {
+                    match &mut aggregated_changeset {
+                        Some(aggregated_changeset) => aggregated_changeset.merge(next_changeset),
+                        aggregated_changeset => *aggregated_changeset = Some(next_changeset),
+                    }
+                    Ok(aggregated_changeset)
                 }
+                Err(iter_error) => Err(AggregateChangesetsError {
+                    changeset: aggregated_changeset,
+                    iter_error,
+                }),
             };
-            match &mut changeset {
-                Some(changeset) => changeset.merge(next_changeset),
-                changeset => *changeset = Some(next_changeset),
-            }
-        }
-        Ok(changeset)
+        // Prepare C changeset deserializer
+        let unversioned_parsed_changeset = self
+            .iter_changesets()
+            .try_fold(Option::<C>::None, changeset_aggregator);
+
+        // But try first versioned V changeset deserializer and if it fails, proceed with C
+        // unversioned deserializer
+        self.iter_versioned_changesets()
+            .try_fold(Option::<C>::None, |acc, x| {
+                changeset_aggregator(acc, x.map(|v| v.into()))
+            })
+            .or(unversioned_parsed_changeset)
     }
```

Comment:
I think there is a bit of a misunderstanding of what @LLFourn is suggesting [here](https://github.com/bitcoindevkit/bdk/issues/1499#issuecomment-2205363144) and [here](https://github.com/bitcoindevkit/bdk/issues/1499#issuecomment-2207582765).

Forward compatibility here means that the encoded `V` is of a newer version than what our software has, but our software needs to support deserializing it.

### To have forward compatibility, we need to prepend a length to every encoded `V`.

Why? Typically, the encoding length can be determined by the type definition of `V`. However, a newer version of `V` will have new variants, and the encoding lengths of those variants cannot be determined (because we don't have it).

### When we fail to deserialize into a variable of type `V`, we need to check the "variant index" encoding first.

Why? If the variant index does not exist in our type `V` (which is an enum), then we are sure the data if of a newer type. Since the data is of a newer type, we can apply the forward-compatibility logic. On the flip side, if the data is of a variant index that that exists in `V`, then we are sure that the data is corrupted.

What is "variant index"? `bincode` encodes enums with the variant index first. The variant index is encoded with `VarintEncoding` by default. Refer to how `bincode` encodes integers [here](https://github.com/bincode-org/bincode/blob/trunk/docs/spec.md#intencoding), and enums [here](https://github.com/bincode-org/bincode/blob/trunk/docs/spec.md#enums).

## The Encoding Format

Each changeset should be encoded as follows:

```
[L][V]
```

* `[L]` is a `bincode` encoded `u64`. Use [`serialized_size`](https://docs.rs/bincode/latest/bincode/config/trait.Options.html#method.serialized_size) on `V` to figure out the value of `L`.
* `[V]` is a `bincode` encoded enum type that represents a versioned changeset. Because it is an enum, it is guaranteed that the encoded data begins with a `VariantEncoding`.

### When encoding a changeset

1. Transform the changeset into a versioned-changeset first.
2. Determine the `serialized_size` of the versioned-changeset. The `serialized_size` of `V` is `L`.
3. Encode `L: u64` and write to file.
4. Enncode `V` and write to file.

### When decoding a changeset

1. Decode `L: u64` from file.
2. Read `L` bytes from file and attempt to decode as `V`.
    * On success: Great! Extract `C` from `V` and return it.
    * On failure: Decode `V` as `u64` (to extract the enum-variant-varint-index). If variant index is too high, trim the variant-index from the data, and attempt to decode into `C` (changeset) directly.

### How do we determine if a variant-index is too high?

Ideally, we can use https://doc.rust-lang.org/std/mem/fn.variant_count.html on our definition of `V`. However, this is a nightly-only API.

Instead, let's use the hack that @LLFourn has suggested: Encode `V::default()` (which should always be the latest-version variant), and then decode the result as `u64` to extract the variant-index.

Reply:
I also propose to update this PR description to: Introduce backwards/forwards compatibility for `bdk_file_store`.

Reply:
Btw, to have forward compatibility, we are assuming that the new changeset `C` only adds new fields at the end as `Option`s (when diffed with the old changeset `C`).

Reply:
Thanks for the thorough review and write up!

Reply:
When decoding a changeset, the `On failure` case for step 2, we hope to succeed if C is of a lower version than the encoded one but without the extra probably `Option` fields or code should fail? Or we expect C to be of the newer version but V doesn't account for it yet? (like in the test I added)

cc @evanlinjin

Reply:
@nymius

Just to be clear with what we are talking about, `V` and `C` can either be type definitions defined by the software version we use to read the data, or by the software version which encoded and wrote the data to the file we are reading.

For type definitions defined by the software that reads, let's use **V<sub>r</sub>** and **C<sub>r</sub>** (Where **C<sub>r</sub>** is the body of **V<sub>r</sub>**'s latest enum variant).

For type definitions defined by the software the writes, let's use **V<sub>w</sub>** and **C<sub>w</sub>** (where **C<sub>w</sub>** is the body of **V<sub>w</sub>**'s latest enum variant).

When decoding the data as **V<sub>r</sub>** fails, it either means:

1. **V<sub>w</sub>**'s enum variant index does not exist in **V<sub>r</sub>**.
2. The data to decode is invalid. This is either data corruption, or **V<sub>r</sub>** has a variant with a different `C` definition to that of **V<sub>w</sub>**.

For `1.`, we still try decode the data (excluding the enum variant index) to **C<sub>r</sub>**. This will work if **C<sub>w</sub>** only adds new fields on **C<sub>r</sub>** (which still exists in **V<sub>w</sub>** as a lower enum variant).

For `2.`, we can just error out. It's either a programming error or data corruption.